packages/adaptive_scaffold/lib/src/slot_layout.dart

```diff
@@ -0,0 +1,188 @@
+
+  @override
+  Widget build(BuildContext context) {
+    chosenWidget = SlotLayout.pickWidget(context, widget.config);
+    bool hasAnimation = false;
+    return AnimatedSwitcher(
+        duration: const Duration(milliseconds: 1000),
+        layoutBuilder: (Widget? currentChild, List<Widget> previousChildren) {
+          final Stack elements = Stack(
+            children: <Widget>[
+              if (hasAnimation) ...previousChildren,
+              if (currentChild != null) currentChild,
+            ],
+          );
+          return elements;
+        },
+        transitionBuilder: (Widget child, Animation<double> animation) {
+          final SlotLayoutConfig configChild = child as SlotLayoutConfig;
+          if (child.key == chosenWidget?.key) {
+            return (configChild.inAnimation != null)
+                ? child.inAnimation!(child, animation)
+                : child;
+          } else {
+            if (configChild.outAnimation != null) {
+              hasAnimation = true;
+            }
+            return (configChild.outAnimation != null)
+                ? child.outAnimation!(child, ReverseAnimation(animation))
+                : child;
+          }
+        },
+        child: chosenWidget ?? SlotLayoutConfig.empty());
+  }
+}
+
+/// Defines how [SlotLayout] should display under a certain [Breakpoint].
+abstract class SlotLayoutConfig extends StatefulWidget {
+  /// Creates a new [SlotLayoutConfig].
+  ///
+  /// Returns the child widget as is but holds properties to be accessed by other
+  /// classes.
+  const SlotLayoutConfig(
+      {required this.builder, this.inAnimation, this.outAnimation, super.key});
+
+  /// The child Widget that [SlotLayout] eventually returns with an animation.
+  final WidgetBuilder? builder;
+
+  /// A function that takes a [Widget] and an [Animation] and returns a [Widget].
+  /// This function is passed to the [AnimatedSwitcher] inside [SlotLayout] and
+  /// is to be played when the child enters.
+  ///
+  /// See also:
+  ///
+  ///  * [AnimatedWidget] and [ImplicitlyAnimatedWidget], which are commonly used
+  ///  as the returned widget.
+  final Widget Function(Widget, Animation<double>)? inAnimation;
+
+  /// A function that takes a [Widget] and an [Animation] and returns a [Widget].
+  /// This function is passed to the [AnimatedSwitcher] inside [SlotLayout] and
+  /// is to be played when the child exits.
+  ///
+  /// See also:
+  ///
+  ///  * [AnimatedWidget] and [ImplicitlyAnimatedWidget], which are commonly used
+  ///  as the returned widget.
+  final Widget Function(Widget, Animation<double>)? outAnimation;
+
+  /// An empty [SlotLayoutConfig] to be placed in a slot to indicate that the slot
+  /// should show nothing.
+  static SlotLayoutConfig empty() {
+    return const _SlotLayoutConfig(key: Key(''), builder: null);
+  }
+
+  @override
+  State<SlotLayoutConfig> createState() => _SlotLayoutConfigState();
+}
+
+class _SlotLayoutConfig extends SlotLayoutConfig {
```

Comment:
If you only wanted to prevent developers constructing a `_SlotLayoutConfig` you can just use `_SlotLayoutConfig._()` constructor. This is a common patter used throughout flutter/flutter, such as https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/mouse_cursor.dart#L359 .

Reply:
But it still bothers me that `SlotLayoutConfig` is a widget. Because while it by itself is a displayable widget, it also contains 2 properties that are not used in the widget at all. In that aspect the widget is also used as a data structure. It feels like [interleaving multiple concept together](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-interleaving-multiple-concepts-together).

And what do you get from making `SlotLayoutConfig` a widget? Nothing much. `_SlotLayoutConfigState.build` is a 3-line long ternary operator, so trivial that the developer can just copy your code or even just come up themselves if they want to "rewrite `SlotLayout`" as you put it. You can even make `builder` non-nullable and require in the documentation whoever overrides it should return `SizedBox.shrink()` if they want it empty. `SlotLayoutConfig` is really a data class. Its widget nature has already been fulfilled by `builder` and `key`.

Reply:
Yes I feel like SlotLayoutConfig would work better as a non-widget class. Do you think there's a better way to handle this?

Reply:
Yeah can you make it a regular class that contains the four properties? Also it doesn't have to be private any more, including the constructor.

Reply:
It would not work even as a class because the AnimatedSwitcher only takes Widgets through the transition builder and the inAnimation/outAnimation properties are needed inside of the transition builder.

Reply:
I see. I think this might be a sign that you should not tie `inAnimation` and `outAnimation` with `SlotLayoutConfig`, because they're not the property of a slot-breakpoint pair, but the property of the entire slot. In other words, what if we move the `inAnimation` and `outAnimation` to `SlotLayout` instead?

Maybe we can make them a function of `Animation<double> (SlotLayoutConfig fromConfig, SlotLayoutConfig toConfig)` so that developers can customize animation based on configs.

Reply:
That could work. It would come at a cost of slightly decreasing customizability but I think it would be worth it.